Atomic.md

```diff
@@ -0,0 +1,190 @@
+<pre>
+Title:       <b>Atomic/Batch Transactions</b>
+Revision:    <b>3</b> (2024-03-14)
+
+Author:      <a href="mailto:mvadari@ripple.com">Mayukha Vadari</a>
+
+Affiliation: <a href="https://ripple.com">Ripple</a>
+</pre>
+
+# Atomic/Batch Transactions
+
+## Abstract
+
+There is a need for transactions to be packaged together and executed atomically.
+
+Potential use cases:
+* Sending an `NFTokenMint` and `NFTokenCreateOffer` transaction in the same fell swoop (if the offer creation fails, so does the mint)
+* Sending a few different offers that might fail and only wanting one to succeed
+* Fees for platforms packaged with the actual transaction
+* Atomic swaps (if multiple accounts are involved)
+
+## 1. Overview
+
+This spec proposes one new transaction: `Atomic`. It will not require any new ledger objects, nor modifications to existing ledger objects. It will require an amendment, tentatively titled `featureAtomic`.
+
+### 1.1. Terminology
+* **Inner transaction**: the sub-transactions included in the `Atomic` transaction, which are the same as the existing set of transactions.
+* **Outer transaction**: the wrapper `Atomic` transaction itself.
+
+## 2. Transaction: `Atomic`
+
+|FieldName | Required? | JSON Type | Internal Type |
+|:---------|:-----------|:---------------|:------------|
+|`TransactionType`|✔️|`string`|`UInt16`|
+|`Account`|✔️|`string`|`STAccount`|
+|`Fee`|✔️|`string`|`STAmount`|
+|`AtomicityType`|✔️|`number`|`UInt8`|
+|`RawTransactions`|!|`array`|`STArray`|
+|`Transactions`|✔️|`array`|`Vector256`|
+|`BatchSigners`| |`array`|`STArray`|
+
+```
+{
+    TransactionType: "Atomic",
+    Account: "r.....",
+    AtomicityType: n,
+    Transactions: [transaction hashes...]
+    RawTransactions: [transaction blobs...], // not included in the signature or stored on ledger
+    BatchSigners: [ // only sign the list of transaction hashes and probably the atomicity type
+      BatchSigner: {
+        Account: "r.....",
+        Signature: "...."
+      },
+      BatchSigner: {
+        Account: "r.....",
+        Signers: [...] // multisign
+      },
+      ...
+    ],
+    SigningPubKey: "....",
+    TxnSignature: "...."
+}
+```
+
+### 2.1. `Fee`
+
+$$(n+2)*base\_fee$$
```

Comment:
I kind of like the simplicity of David S's fee design. WDYT?

> ...only the outer transaction pays a fee. The fee is equal to the sum of the fees required for all inner transactions plus the cost of the outer transaction itself. The fee is claimed by the outer transaction's execution and no inner transactions claim fees.

Transactions fees should be De minimis in general (so outer account pays all tx fees shouldn't be too much of a burden), but the bigger advantage I see in just having one entity pay the fees is that it allows only a single entity to try and determine what fees might need to be to  make the transaction succeed, and this can then be done only once -- at the time of submission, which is arguably the best time to decide how much will need to be paid.

Reply:
It just seems a bit complicated to implement that.

Reply:
Actually, after some further digging, I think as long as we don't check `calculateBaseFee` in the inner transaction processing we should be good.

Reply:
I don't remember if we discussed this directly, but I like this change because then fee escalation applies on the total fee, not on just the batch processing fee.